projects/com.oracle.truffle.llvm.parser.bc.impl/src/com/oracle/truffle/llvm/parser/bc/impl/LLVMMetadata.java

```diff
@@ -220,17 +222,84 @@ public void visit(GetElementPointerInstruction gep) {
                 StructureType thisStruct = (StructureType) t1;
                 // TODO: should always be this type?
                 if (thisStruct.getMetadataReference().isPresent()) {
-                    MetadataCompositeType metaStruct = (MetadataCompositeType) thisStruct.getMetadataReference().get();
-                    thisStruct.setName(((MetadataString) metaStruct.getName().get()).getString());
-
-                    MetadataNode elements = (MetadataNode) metaStruct.getMemberDescriptors().get();
-
-                    Symbol idx = gep.getIndices().get(1);
-                    int parsedIndex = idx instanceof IntegerConstant ? (int) ((IntegerConstant) (idx)).getValue() : 0;
-                    MetadataReference element = elements.get(parsedIndex);
-
-                    MetadataDerivedType derivedType = (MetadataDerivedType) element.get();
-                    gep.setReferenceName(((MetadataString) derivedType.getName().get()).getString());
+                    MetadataBaseNode node = thisStruct.getMetadataReference().get();
+
+                    // MetadataCompositeType
+                    // MetadataDerivedType
+                    if (node instanceof MetadataCompositeType) {
+                        MetadataCompositeType metaStruct = (MetadataCompositeType) node;
+                        thisStruct.setName(((MetadataString) metaStruct.getName().get()).getString());
+
+                        MetadataNode elements = (MetadataNode) metaStruct.getMemberDescriptors().get();
+
+                        Symbol idx = gep.getIndices().get(1);
+                        int parsedIndex = idx instanceof IntegerConstant ? (int) ((IntegerConstant) (idx)).getValue() : 0;
+
+                        long elementOffset = thisStruct.getElementOffset(parsedIndex);
+                        for (MetadataReference element : elements) {
+                            MetadataDerivedType derivedType = (MetadataDerivedType) element.get();
+                            if (derivedType.getOffset() == elementOffset) {
+                                gep.setReferenceName(((MetadataString) derivedType.getName().get()).getString());
+                                break;
+                            }
+                        }
+                    } else if (node instanceof MetadataBasicType) {
+                        // TODO: implement?
+                    } else if (node instanceof MetadataDerivedType) {
+                        // TODO: implement?
+                    } else {
+                        throw new AssertionError("unknow node type: " + node);
+                    }
+                }
+            } else {
+                if (gep.getBasePointer() instanceof CastInstruction) {
```

Comment:
The code nesting is really deep in this method. Could you extract some logic into methods?